feat(@embark/embark-rpc-manager): Add support for `eth_signTypedData_v3`

emizzle opened the pull request:
This PR adds support for EIP-712 (https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md), by allowing the signing of typed data transactions using the `eth_signTypeData_v3` or `eth_signTypedData` request.

Add a module called `embark-rpc-manager` to handle all RPC transaction modifications.

Add a module called `embark-events` which is the internal events management module extracted. This allows other modules (ie plugins) to use a private instance of Embark’s event API.

Remove transaction modifications in `embark-accounts-manager` and leave only the functionality in this module that handle account management (ie funding of accounts). Other functionality was moved to `embark-rpc-manager`.

Support for custom EmbarkJS plugins.

### TODO
- The RPC error codes passed back from the Proxy are basically hardcoded to `-32603` -- `Internal JSON-RPC error` (according [to the JSON-RPC spec](https://www.jsonrpc.org/specification)) -- regardless of the source of error. We should change these error codes to be more accurate instead of returning the same error code for any error.

### Other changes
- Transactions should now reflect when a new node account is added via `personal_newAccount` or via the node itself (and once `eth_accounts` is requested).
- In the proxy, errors are handled for all cases and are now JSON-RPC spec-compliant.
- Always register `eth_signTypedData` RPC response and display error message if account cannot be found.

### NOTE
- This PR is based on #1983 and should be rebased on to master when (and if) #1983 is merged.

jrainville commented on packages/embark-proxy/src/index.ts:
```diff
@@ -71,14 +71,24 @@ export default class ProxyManager {
+        !this.isWs ? this.wsPort : this.rpcPort,
+        !this.isWs,
+        null,
+      );
```
This whole section is done a bit wrong.
It should be something like:
```
// HTTP
this.proxy = await new Proxy({ events: this.events, plugins: this.plugins, logger: this.logger })
      .serve(
        this.embark.config.blockchainConfig.endpoint,
        this.host,
        this.rpcPort,
        false,
        null,
      );
if (this.ws){
this.proxy = await new Proxy({ events: this.events, plugins: this.plugins, logger: this.logger })
      .serve(
        this.embark.config.blockchainConfig.endpoint,
        this.host,
        this.wsPort,
        true,
        null,
      );
}
```

emizzle replied:
whoops! totally forgot to remove that after testing. I had no way of testing WebSockets, so I started an HTTP proxy and meant to remove it. Removed.

jrainville replied:
I still see the double Proxy. Didn't you say you removed? Is it a Github bug?

emizzle commented on packages/embark-transaction-manager/src/lib/eth_accounts.ts:
```diff
@@ -0,0 +1,36 @@
+    const txsToModify = [
+      blockchainConstants.transactionMethods.eth_accounts,
+      blockchainConstants.transactionMethods.personal_listAccounts,
+    ];
```
Good idea.

emizzle commented on packages/embark-transaction-manager/src/lib/eth_accounts.ts:
```diff
@@ -0,0 +1,36 @@
+      blockchainConstants.transactionMethods.personal_listAccounts,
+    ];
+
+    if (!(txsToModify.some((txMethod) => txMethod === params.reqData.method) && this.accounts.length)) {
```
Great idea. Fixed.

emizzle commented on packages/embark-transaction-manager/src/lib/index.ts:
```diff
@@ -0,0 +1,43 @@
+      if (!this._accounts) {
+        const web3 = await this.web3;
+        const nodeAccounts = await web3.eth.getAccounts();
+        this._accounts = AccountParser.parseAccountsConfig(this.embark.config.blockchainConfig.accounts, web3, dappPath(), this.logger, nodeAccounts);
```
Accounts manager = funding
Transaction manager = modify transactions

emizzle replied:
Good point. Should I re-merge them? Just feels a bit odd to be included tx modifications in an accounts manager...

jrainville replied:
Well, I did the original accounts manager, so I liked it all in the same place. We could try to get the opinion of someone else to see what to  keep. 
The only thing negative I see with what you have right now is that the accounts-manager is only funding and we do the account parsing twice.

emizzle replied:
I changed the base class name for each method modification to `RpcModifier`. And then I changed the package name to `embark-rpc-manager`. What do you think about this name?

I agree that parsing the accounts twice is not great, but I don't think it's a huge performance hit, and it allows us to separate these two plugins for their individual concerns (ie embark-account-manager` handles funding or possibly future account operations, and `embark-rpc-manager` handles modification of RPC calls to/from Embark).

emizzle commented on packages/embark-transaction-manager/src/lib/eth_sendTransaction.ts:
```diff
@@ -0,0 +1,80 @@
+    });
+  }
+  private async checkRequestFor_eth_sendTransaction(params: any, callback: Callback<any>) {
+    if (!(params.reqData.method === blockchainConstants.transactionMethods.eth_sendTransaction && this.accounts.length)) {
```
Fixed.

emizzle commented on packages/embark-transaction-manager/src/lib/eth_signTypedData.ts:
```diff
@@ -0,0 +1,45 @@
+    const txsToModify = [
+      blockchainConstants.transactionMethods.eth_signTypedData_v3,
+      blockchainConstants.transactionMethods.eth_signTypedData,
+    ];
```
Fixed.

emizzle commented on packages/embark-transaction-manager/src/lib/eth_signTypedData.ts:
```diff
@@ -0,0 +1,45 @@
+      blockchainConstants.transactionMethods.eth_signTypedData,
+    ];
+
+    if (!txsToModify.some((txMethod) => txMethod === params.reqData.method)) {
```
Updated.

emizzle commented on packages/embark-transaction-manager/src/lib/eth_accounts.ts:
```diff
@@ -0,0 +1,36 @@
+    this.logger.trace(__(`Original request/response data: ${JSON.stringify(params)}`));
+
+    try {
+      params.respData.result = this.accounts.map((acc) => acc.address || acc);
```
Great point. I didn't actually think of node accounts changing. I added a tx interception for `personal_newAccount`, which then informs the transaction-manager (via a private event bus), and all the transaction modifiers will have their account modified accordingly. This will handle cases when a user adds a new account via `web3.personal.newAccount` API. 

For the lines highlighted above, it seems this would handle the case if the user adds a node account to the node itself, and then a call is made to `eth_accounts`. I added back some of the code to modify the response in that case, but also emits an event so that all other tx modifiers can have their node accounts updated as well.

emizzle commented on packages/embark-transaction-manager/src/lib/eth_signTypedData.ts:
```diff
@@ -0,0 +1,45 @@
+import Web3 from "web3";
+const { blockchain: blockchainConstants } = require("embark-core/constants");
+
+export default class EthSignTypedData {
```
Because geth and most other nodes don't support it (yet or ever). Instead they return the error "The method eth_signTypedData_v3 does not exist/is not available".

jrainville commented on packages/embark-transaction-manager/src/lib/eth_accounts.ts:
```diff
@@ -0,0 +1,61 @@
+        // reset backing variables so accounts is recalculated
+        this.txEvents.emit("nodeAccounts:updated", params.respData.result);
+        // update accounts for this response only.
+        accounts = AccountParser.parseAccountsConfig(this.embark.config.blockchainConfig.accounts, this.web3, dappPath(), this.logger, nodeAccounts);
```
Those two lines could be replaced by a request instead of an emit, because then, the tx-manager would have time to modify the accounts, then this flow can continue and `this.accounts` are the new accounts (so no need to parse twice)

emizzle replied:
Great idea. Updated.

jrainville commented on packages/embark-transaction-manager/src/lib/eth_accounts.ts:
```diff
@@ -0,0 +1,61 @@
+
+  private async checkResponseFor_eth_accounts(params: any, callback: Callback<any>) {
+
+    if (!(TXS_TO_MODIFY.includes(params.reqData.method))) {
```
Jut realized, `TXS_TO_MODIFY` is not really the right word, it should be more `METHODS_TO_MODIFY`, because `eth_accounts` and the other are not txs.

emizzle replied:
Updated.

jrainville commented on packages/embark-transaction-manager/src/lib/eth_sendTransaction.ts:
```diff
@@ -0,0 +1,87 @@
+    // Allow to run transaction in parallel by resolving the nonce manually.
+    // For each transaction, resolve the nonce by taking the max of current transaction count and the cache we keep locally.
+    // Update the nonce and sign it
+    this.signTransactionQueue = async.queue(({ payload, account }, callback: Callback<any>) => {
```
Might as well include the transactionQueue in the `if` or reverse the `if` and return?

emizzle replied:
Updated.

jrainville commented on packages/embark-transaction-manager/src/lib/eth_signTypedData.ts:
```diff
@@ -0,0 +1,53 @@
+import { EventEmitter } from "events";
+import Transaction from "./transaction";
+
+const TXS_TO_MODIFY = [
```
Again, this is more methods and not TXs

emizzle replied:
Updated, thanks.

jrainville commented on packages/embark-transaction-manager/src/lib/index.ts:
```diff
@@ -0,0 +1,85 @@
+        modifier.accounts = accts;
+      }
+    });
+  }
```
The two events above almsot have the same function content. Might be nice to find a way to merge them
Something like:
```
reprocessAccounts(updatedNodeAccounts) {
  this._nodeAccounts = updatedNodeAccounts;
      this._accounts = null;
      const accts = await this.accounts;
      for (const modifier of this.modifiers) {
        modifier.accounts = accts;
      }
}

this.txEvents.on("nodeAccounts:updated", reprocessAccounts.bind(this));
this.txEvents.on("nodeAccounts:added", async (addedNodeAccount) => {
    if (!this._nodeAccounts) {
      this._nodeAccounts = [addedNodeAccount];
    } else {
      this._nodeAccounts.push(addedNodeAccount);
    }
  reprocessAccounts(this._nodeAccounts);
});
```

emizzle replied:
Updated.

jrainville commented on packages/embark-transaction-manager/src/lib/personal_newAccount.ts:
```diff
@@ -0,0 +1,25 @@
+    }
+
+    // emit event so tx modifiers can refresh accounts
+    this.txEvents.emit("nodeAccounts:added", params.respData.result);
```
You need to call the callback here

emizzle replied:
Updated, thanks.

iurimatias commented on packages/embark-events/README.md:
```diff
@@ -0,0 +1,7 @@
+embark-events
```
what's is the reasoning for creating a package for this? the events are integral part of thee core api, IMO they shouldn't another package, we have a bit too many packages already.

emizzle replied:
The reasoning is so that we can use a private instance embark events inside of another package `embark-rpc-manager`, so as not to pollute the global events (which could incorrectly allow other developers to use events they're not meant to be using - and potentially cause race conditions).

iurimatias commented on packages/embark-rpc-manager/src/lib/index.ts:
```diff
@@ -0,0 +1,84 @@
+    const nodeAccounts = await this.nodeAccounts;
+    const accounts = await this.accounts;
+
+    this.rpcModifierEvents.setCommandHandler("nodeAccounts:updated", this.updateAccounts.bind(this));
```
feels unnecessary, why not just usee the main event bus properly namespaced?

iurimatias replied:
see embark-deployment, it also uses event bus, but doesn't need a new one, just namespaced

emizzle replied:
The reasoning is so that we can use a private instance embark events inside of another package embark-rpc-manager, so as not to pollute the global events (which could incorrectly allow other developers to use events they're not meant to be using - and potentially cause race conditions).

michaelsbradleyjr commented on packages/embark-accounts-manager/src/index.ts:
```diff
@@ -64,91 +28,34 @@ export default class AccountsManager {
           return fundAccount(web3, account.address, coinbase, account.hexBalance);
       });
-      await Promise.all(fundingAccounts);
+      await Promise.all([fundingAccounts]);
```
```suggestion
      await Promise.all(fundingAccounts);
```

`fundingAccounts` is already an iterable so I it shouldn't wrapped in `[]`. When it's wrapped the Promise will resolve on the next turn of the microtask queue with an array of promises that will likely not have resolved yet.

michaelsbradleyjr replied:
Try this in a Node.js REPL:

```js
p = () => new Promise(resolve => { setTimeout(resolve, 3000) })

x = [p(), p()]; y = Promise.all([x]).then(arr => console.log(arr))
```
You'll see printed:
```
[ [ Promise { <pending>, domain: [Domain] },
    Promise { <pending>, domain: [Domain] } ] ]
```
so the Promise.all resolved on the next turn of the microtask queue, i.e. almost immediately, but the promises in the wrapped array haven't resolved.

compare with:
```js
x = [p(), p()]; y = Promise.all(x).then(arr => console.log(arr))
```
only after ~3 seconds this will be printed:
```
[ undefined, undefined ]
```

emizzle replied:
Thanks for the explanation @michaelsbradleyjr. I fixed this by removing the change and adding one additional null check and the typechecker seems to be happy now (the typechecker was happy without the additional null check as well)... 🤷‍♂️

michaelsbradleyjr commented on packages/embark-events/package.json:
```diff
@@ -0,0 +1,68 @@
+    "npm": ">=6.4.1",
+    "yarn": ">=1.12.3"
+  },
+  "dependencies": {
```
Indeed.

emizzle replied:
Updated, thank you for that!

emizzle commented:
@iurimatias - i don't agree that we should remove `embark-events`. If we use the main event bus, we are "crowding" the global event space with events that are not meant to be consumed by other packages/plugins. This opens us up to other developers using these events incorrectly -- 
a breeding for potential race conditions. Allowing packages to create a private instance of Embark's event bus prevents this issue, while still allowing us to take advantage of `.request/2` inside of Embark's event bus.

iurimatias commented:
@emizzle any updates on this?

emizzle commented:
Updated the following:
- Rebased
- Moved `embark-events` to `packages/core/core` (`embark-core`)
- Moved `packages/embark-rpc-manager` to `packages/plugins/rpc-manager` and updated dependent embark packages to version `4.1.1`.\

I left the typings on `packages/core/typings/src/embark.d.ts` and duplicated them in `packages/core/core/` (`embark-core`) to avoid having to update all packages to depend on `embark-core`.

There are typecheck issues on the package `embark-rpc-manager` (lives at `packages/plugins/rpc-manager`. When I do `yarn typecheck`, I get errors relating to cockpit dependencies:
```
$ tsc
../../../node_modules/connected-react-router/index.d.ts:2:26 - error TS7016: Could not find a declaration file for module 'react'. '/Users/emizzle/Code/__Github/embk-fw/embark/node_modules/react/index.js' implicitly has an 'any' type.
  Try `npm install @types/react` if it exists or add a new declaration (.d.ts) file containing `declare module 'react';`

2   import * as React from 'react'
                           ~~~~~~~

../../../node_modules/connected-react-router/index.d.ts:4:84 - error TS7016: Could not find a declaration file for module 'history'. '/Users/emizzle/Code/__Github/embk-fw/embark/node_modules/history/index.js' implicitly has an 'any' type.
  Try `npm install @types/history` if it exists or add a new declaration (.d.ts) file containing `declare module 'history';`

4   import { History, Path, Location, LocationState, LocationDescriptorObject } from 'history'
                
```

jrainville commented on packages/embarkjs/build-test/blockchain.js.map:
```diff
@@ -0,0 +1 @@
+{"version":3,"sources":["../test/blockchain.js"],"names":["startRPCMockServer","TestProvider","require","assert","Blockchain","promisify","describe","before","registerProvider","setProvider","scenarios","description","servers","visited","error","forEach","scenario","it","makeServers","all","map","server","successful","dappConnection","connectionString","resolve","reject","connect","err","idx","st ...
```
You have some `.map`  files that snucked their way in

emizzle replied:
Fixed, thanks!

emizzle commented:
Found a race condition where RPC calls are being executed before registration of the RPC modifiers. Currently working to resolve this. Please hold off merging until that's done.

emizzle commented:
@jrainville - this is now loading an HTTP proxy in always (as well as a WS proxy if running over WS), so as to allow connections from MetaMask.

I have also removed all `accounts` checks in the constructors of the RPC modifications as this was causing a race condition (`web3` needs to be determined at time of request, and not after blockchain is started, otherwise RPC calls will not be registered for interception until it's too late and some RPC calls will get the unmodified response).

emizzle commented on packages/stack/proxy/src/index.ts:
```diff
@@ -81,18 +82,30 @@ export default class ProxyManager {
+    // HTTP
+    this.httpProxy = await new Proxy({ events: this.events, plugins: this.plugins, logger: this.logger })
+      .serve(
+        this.embark.config.blockchainConfig.endpoint,
```
@jrainville Sorry, this must have been missed in the rebase. 

Updated - only starting WS proxy in the case of VM.

emizzle commented:
Rebased with master

PascalPrecht commented on packages/stack/proxy/src/proxy.js:
```diff
@@ -63,7 +63,7 @@ export class Proxy {
 
     if (this.isWs) {
-      this.app.ws('/', async (conn, _wsReq) => {
+      this.app.ws('/', async (ws, _wsReq) => {
```
This is renamed to `ws` but the code still relies on `conn`. See L:68 and L:76

PascalPrecht commented:
Unfortunately, after merging #2016, this has introduced a lot of conflicts.

@emizzle I'll take care of rebasing this one here again. Sorry for the inconvenience.

PascalPrecht commented:
@emizzle rebased this one on latest `master`. Just to be sure, double check with comment: https://github.com/embark-framework/embark/pull/1850#discussion_r344689490

emizzle commented on packages/core/core/index.d.ts:
```diff
@@ -1,5 +1,32 @@
+    opt10?: any,
+    opt11?: any,
+    opt12?: any,
+  ) => any;
```
I wholeheartedly agree with this, I simply moved the code from another place and didn't want to also change it, as I thought it might cloud the purpose of the PR.

PascalPrecht commented:
> The changes in #1850 (review) still need to be addressed.

Addressing these now

PascalPrecht commented on packages/plugins/accounts-manager/src/index.ts:
```diff
@@ -73,84 +34,27 @@ export default class AccountsManager {
-      const accts = this.accounts
-        .filter((account) => account.address);
+      const accts = accounts
+        .filter((account) => account && account.address);
```
This is the line that makes things blow up right now with

```
Error returned from the node: Returned error: insufficient funds for gas * price + value
```

After some debugging I found out that the `account` used to deploy the contracts has zero funds. The reason for that is that in this line here, we're filtering out all accounts that don't have an `address`.

Unfortunately, when this code is run, none of the accounts will have an `address`, rather, they are addresses themselves. The reason for that is because of how we changed the retrieval of accounts [here](https://github.com/embark-framework/embark/pull/1850/commits/9bea7121d6e54ed8f079a249690364aef5a9b735#diff-cff683ca704c06d26b59ecd676319e87R41). 

E.g. we're relying purely on `web3.eth.getAccounts()` which will always return a list of strings (addresses), therefore, this filtering here causes ground zero.

This used to work before, when we were relying on `AccountParser.parseAccountsConfig()`, which we now also do [here](https://github.com/embark-framework/embark/pull/1850/commits/9bea7121d6e54ed8f079a249690364aef5a9b735#diff-9c48cf3ec968937697f7de95f2410e83R49).

To make this work, we can simply remove this filter and change `acct.address` to `acct` [here](https://github.com/embark-framework/embark/pull/1850/commits/9bea7121d6e54ed8f079a249690364aef5a9b735#diff-cff683ca704c06d26b59ecd676319e87L156), **however** at this point I'm not sure if we indeed really just want to rely on `web3.eth.getAccounts()` or if we still need/want to consider `parseAccountsConfig()`.

Thoughts @emizzle ?

PascalPrecht commented:
Updated this PR:

- rebased on top of latest `master`
- addressed @michaelsbradleyjr 's comments in `680e01c`
- applied changes to fix bug discussed in https://github.com/embark-framework/embark/pull/1850#discussion_r346255759 (needs review by @emizzle ) - also done in `680e01c` for easier reviewing **but needs squashing once signed off**

emizzle commented:
LGTM. I think this was partly a bad rebase on my part. Let's get @jrainville's input as I believe the `eachLimit` was introduced as a replacement for `Promise.all`.

jrainville commented on packages/plugins/accounts-manager/src/index.ts:
```diff
@@ -73,87 +34,28 @@ export default class AccountsManager {
-      async.eachLimit(accts, 1, (acct, eachCb) => {
-        fundAccount(web3, acct.address, coinbase, acct.hexBalance)
+      async.eachLimit(accounts, 1, (acct, eachCb) => {
+        fundAccount(web3, acct, coinbase)
```
As @PascalPrecht said, we used to rely on accounts returned from `parseAccountsConfig`, but now we only get the addresses. That's a problem, because we no longer fund the accounts up to the balance that they require. This will cause some tests to fail.

I don't remember the reason why the parsing was moved to another module, but if this module is the accounts manager, I would expect it to be the source of the accounts, now it's more like the "account funder", and even at that, it doesn't perform correctly.

Maybe we move back the parsing here and have the other modules get the accounts from here (but it's a plugin, so probably doesn't work), or we also put back the RPC call changes.
We can also have this module get the accounts from the other, but I think the other is a plugin too.

emizzle replied:
So with the `rpc-manager` in place and the proxy working correctly, a call to `web3.eth.getAccounts()` should get the correct accounts. No need to have a module being the source of truth.

jrainville commented on packages/plugins/accounts-manager/src/index.ts:
```diff
@@ -1,66 +1,27 @@
       this._web3 = null;
-      this.parseAndFundAccounts(null);
-    });
-    this.embark.registerActionForEvent("accounts:reseted", async (params, cb) => {
```
As described in Discord, this is still needed to reset the accounts during the tests.

emizzle replied:
Bad rebase, my fault. Need to put this back in.

jrainville commented on packages/core/core/src/plugin.ts:
```diff
@@ -125,6 +125,10 @@ export class Plugin {
       this.acceptedContext = [this.acceptedContext];
     }
+    this.registerActionForEvent("tests:config:updated", { priority: 30 }, async ({ accounts }, cb) => {
+      this.config.blockchainConfig.accounts = accounts;
```
This should already ne done by a `request`, but that's fine too, more direct

PascalPrecht commented on packages/plugins/coverage/src/lib/index.ts:
```diff
@@ -25,7 +25,7 @@ export default class Coverage {
     }
 
-    this.embark.registerActionForEvent("tests:contracts:compile:before", async (contractsFiles, cb) => {
+    this.embark.registerActionForEvent("tests:contracts:compile:before", undefined, async (contractsFiles, cb) => {
```
I might be missing something here, but this seems to be a change we don't want? In fact, it looks like this will turn off the actual action here.

emizzle replied:
It's required for TS to be happy since we previously updated the signature to accept a `{ priority: integer }` option

PascalPrecht replied:
Okay, as discussed offline, this parameter was introduced due to the new `priority` API introduced in https://github.com/embark-framework/embark/commit/776db1b7f71e9a78f216cf2acc6c1387c60b3604

However, there's a bug in our typings that forces users to do this.

Sent a fix here: https://github.com/embark-framework/embark/pull/2060

Once that lands, the `undefined` can be removed here

/cc @emizzle

PascalPrecht commented:
Rebased this on top of latest master, therefor addressing https://github.com/embark-framework/embark/pull/1850#discussion_r347299334

Once CI is green, this should (finally) be ready to go in. 

Huge kudos to @emizzle for the effort and patience, once again!
